tests/test_decorators.py

```diff
@@ -316,11 +340,21 @@ def validate_bar(self, value):
         assert '"bar" field does not exist.' in str(excinfo)
 
     def test_precedence(self):
-        class Schema2(ValidatesSchema):
+        class Schema2Base(Schema):
+            foo = fields.Int()
+
+            @validates('foo')
```

Comment:
This is an interesting/tricky edge case.

I can't actually use the original `ValidatesSchema` here, because `@validates(foo)` tries to validate the `foo` field declared on `ValidatesSchema`.

However, that field gets overridden in `Schema2`, and thus can't be found when trying to validate it.

Reply:
So if I'm understanding this correctly, `@validates('foo')` and `@validates(foo)` behave differently:

- `@validates(foo)` will always validate the `foo` field instance passed in
- `@validates('foo')` searches for `foo` according to the MRO.

So users would have to be careful when doing `@validates(foo)` on base schema classes because subclasses won't be able to override `foo`.

Reply:
yes. thinking about it some more, this is pretty weird DX-wise.

a different way to do this might be to bind the fields to their names in `SchemaMeta`.

that way, assuming the field has been bound to its name, then we can look up the name of the field from the field object, and not have to care about the MRO. it would also let us get rid of much of the extra code here around building the extra mapping.

the downside, then, is that this would limit the re-use of field objects. so for example, right now i believe this is okay:

```python
field = fields.Integer()

class FooSchema(Schema):
    foo = field

class BarSchema(Schema):
    bar = field
```

the straightforward way to do this (binding the field object to the assigned name in `SchemaMeta`) would prohibit the above, though it would still allow

```python
field = fields.Integer()

class FooSchema(Schema):
    foo = field

class BarSchema(Schema):
    foo = field
```

Reply:
@sloria Any thoughts on the above?